File: unity/Assets/Scripts/Multiplayer/Rooms/RoomShort.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace Michitai.Multiplayer.Rooms
{
    /// <summary>
    /// Minimal information about a game room.
    /// Contains basic room details without full player information.
    /// Uses Unity's JsonUtility for deserialization of rules.
    /// </summary>
    /// <typeparam name="T">The type to deserialize room rules into.</typeparam>
    [System.Serializable]
    public class RoomShort<T> where T : class, new()
    {
        [SerializeField]
        private string rules_json;   // Unity mode

        /// <summary>
        /// The unique ID of the room.
        /// </summary>
        public string room_id;

        /// <summary>
        /// The name of the room.
        /// </summary>
        public string room_name;

        /// <summary>
        /// Maximum number of players allowed.
        /// </summary>
        public int max_players;

        /// <summary>
        /// Current number of players in the room.
        /// </summary>
        public int current_players;

        /// <summary>
        /// Whether the room has a password set.
        /// </summary>
        public bool has_password;

        /// <summary>
        /// Whether host switching is allowed.
        /// </summary>
        public bool host_switch;

        /// <summary>
        /// Whether players can leave the room.
        /// </summary>
        public bool can_leave;

        /// <summary>
        /// Whether the room supports realtime communication.
        /// </summary>
        public bool realtime;

        /// <summary>
        /// The room rules deserialized into the specified type.
        /// </summary>
        public T Rules
        {
            get
            {
                return JsonUtility.FromJson<T>(rules_json);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/RoomShort.java
package com.michitai.multiplayer.rooms;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Short room information for room listings.
 *
 * @param <T> The type to deserialize room rules into.
 */
public class RoomShort<T> {
    @JsonProperty("room_id")
    private String roomId;

    @JsonProperty("room_name")
    private String roomName;

    @JsonProperty("current_players")
    private int currentPlayers;

    @JsonProperty("max_players")
    private int maxPlayers;

    @JsonProperty("has_password")
    private boolean hasPassword;

    @JsonProperty("host_switch")
    private boolean hostSwitch;

    @JsonProperty("can_leave")
    private boolean canLeave;

    @JsonProperty("realtime")
    private boolean realtime;

    @JsonProperty("rules_json")
    private String rulesJson;

    @JsonProperty("rules")
    private T rules;

    public String getRoomId() {
        return roomId;
    }

    public void setRoomId(String roomId) {
        this.roomId = roomId;
    }

    public String getRoomName() {
        return roomName;
    }

    public void setRoomName(String roomName) {
        this.roomName = roomName;
    }

    public int getCurrentPlayers() {
        return currentPlayers;
    }

    public void setCurrentPlayers(int currentPlayers) {
        this.currentPlayers = currentPlayers;
    }

    public int getMaxPlayers() {
        return maxPlayers;
    }

    public void setMaxPlayers(int maxPlayers) {
        this.maxPlayers = maxPlayers;
    }

    public boolean isHasPassword() {
        return hasPassword;
    }

    public void setHasPassword(boolean hasPassword) {
        this.hasPassword = hasPassword;
    }

    public boolean isHostSwitch() {
        return hostSwitch;
    }

    public void setHostSwitch(boolean hostSwitch) {
        this.hostSwitch = hostSwitch;
    }

    public boolean isCanLeave() {
        return canLeave;
    }

    public void setCanLeave(boolean canLeave) {
        this.canLeave = canLeave;
    }

    public boolean isRealtime() {
        return realtime;
    }

    public void setRealtime(boolean realtime) {
        this.realtime = realtime;
    }

    public String getRulesJson() {
        return rulesJson;
    }

    public void setRulesJson(String rulesJson) {
        this.rulesJson = rulesJson;
    }

    public T getRules() {
        return rules;
    }

    public void setRules(T rules) {
        this.rules = rules;
    }
}
